### PR TITLE
Add e2e sanity tests

### LINKED ISSUE #40: KFP Testing
We should make sure that KFP is working well. Check if any of the component of KFP (DB, storage, creds etc.) can we replaced by AWS services from integration POV. We should also point any standard example from official doc which has been tested.

### PR DESCRIPTION
**Pull request by rrrkharse**
**Which issue is resolved by this Pull Request:**
Resolves #40 #30 

**Description of your changes:**
Adding sanity tests that install the vanilla distribution of kubeflow and verify that KFP and Katib are working as intended.
This PR will also lay the groundwork for future e2e tests to be implemented upon (e.g. framework implementation).

Todo:

- [x] KFP sanity test
- [x] Katib sanity test
- [x] Usable README
- [x] Misc. todos in file comments

Testing:
```
kf-e2e-test-env ❯ pytest -s -q --region ap-south-1
...

3 passed in 3477.65s (0:57:57)

~/kf-e2e-tests/e2e v1.4-branch-e2e-testing* 57m 58s
```

Also ran with `--metadata` and `--keepsuccess` during test development.

### CODE REVIEW

File: e2e/test_pipelines.py [hunk 1]
```diff
@@ -0,0 +1,276 @@
+from os import access
```
Comment (rrrkharse) [inline e2e/test_pipelines.py:1]: This file is incomplete and does not belong in this PR

File: e2e/README.md [hunk 2]
```diff
@@ -0,0 +1,41 @@
+
+## Usage
+
+Install dependencies
+```
+python3
+kustomize   # MUST BE VERSION 3.2.0
```
Comment (goswamig) [inline e2e/README.md:7]: this has to be precise this or >=3.2.0 ?

File: e2e/README.md [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+Run without deleting successfully created resources. 
+Usefull for rerunning failed tests.
+```
+pytest -s -q --keepsuccess --region <REGION_NAME>
+```
+
+Resume from a previous run using the resources that were previous created
+```
+pytest -s pytest -s -q --metadata .metadata/metadata-1638939746471968000 --keepsuccess --region <REGION_NAME>
```
Comment (goswamig) [inline e2e/README.md:40]: whats the number `1638939746471968000` ?

File: e2e/config.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
+import pytest
```
Comment (goswamig) [inline e2e/config.py:1]: some comments over file on what the intended action will help future reader to understand the code.

File: e2e/README.md (hunk 2 above)
Comment (rrrkharse) [reply e2e/README.md:7]: Must be precise due to changes in kustomize from newer versions that affects resource creation order. This needs to be updated in the other READMEs as well, will create an issue.

File: e2e/README.md (hunk 3 above)
Comment (rrrkharse) [reply e2e/README.md:40]: Time in nanoseconds

File: e2e/config.py (hunk 4 above)
Comment (mbaijal) [reply e2e/config.py:1]: Agreed, add [docstrings](https://www.python.org/dev/peps/pep-0257/#what-is-a-docstring) to functions specially those in the utility files such as `cluster.py` and `config.py`.

File: e2e/README.md (hunk 2 above)
Comment (goswamig) [reply e2e/README.md:7]: If it has to be updated in README for customer please make it as p0 @surajkota @mbaijal fyi.

File: e2e/README.md (hunk 3 above)
Comment (goswamig) [reply e2e/README.md:40]: I believe that file (`meta-xxxxxxxxx`) is being created when u run it right ?

File: e2e/test_pipelines.py (hunk 1 above)
Comment (goswamig) [reply e2e/test_pipelines.py:1]: can u please remove it then

**Review by goswamig — APPROVED** [review]
Good work !!!

Few minor comments, I think we should get going with comments.

File: e2e/README.md (hunk 3 above)
Comment (rrrkharse) [reply e2e/README.md:40]: Update the README on this:

### About metadata
When using the helper method `configure_resource_fixture` a metadata file is generated with the following output:
```
# Using cluster as an example

Saved key: cluster_name value: e2e-test-cluster-507uvuyhca in metadata file /Users/rkharse/kf-e2e-tests/e2e/.metadata/metadata-1639087995874492000
```

Metadata can also be manually be output to a file by calling `Metadata#to_file` on a metadata object

So while the test is running, once it has created a resource it will create a new metadata file that will capture the state of all resources created in the test so far.

If a test fails for some reason, and the `--keepsuccess` flag is present then no resources will be deleted.

Then the next test run can be given the last metadata file via `--metadata <FILE>` and the test will use the previously created resources and skip creating them. 

This can save a lot of time (e.g. avoid having to create a new cluster, install kubeflow for every run).

**Comment by rrrkharse** [comment]
Ok made the changes, am ready for review.

File: e2e/kustomize.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    Before all tests are run, installs kubeflow using the manifest at `kustomize_path` 
+    if `kustomize_path` was not provided in the metadata.
+
+    After all tests are run, uninstalls kubeflow using the manifest at `kustomize_path` 
+    if the flag `--keepsuccess` was not provided as a pytest argument.
+    """
+
+    def on_create():
+        wait_for(lambda : apply_kustomize(kustomize_path), timeout=20*60)
+        time.sleep(5*60)    # wait a bit for all pods to be running
```
Comment (ryansteakley) [inline e2e/kustomize.py:56]: Is there any basis for choosing this amount of time?

Comment (rrrkharse) [reply e2e/kustomize.py:56]: I usually see all pods are running after 5 minutes

**Review by goswamig — APPROVED** [review]

**Review by surajkota — CHANGES_REQUESTED** [review]
nice work, few comments/questions

File: e2e/cluster.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+import subprocess
+import pytest
+
+from e2e.utils import rand_name
+from e2e.config import configure_resource_fixture
+
+def create_cluster(cluster_name, region, cluster_version='1.19'):
+    cmd = []
+    cmd += "eksctl create cluster".split()
```
[the remark this answers is not in the document]
Comment (surajkota) [reply e2e/cluster.py:13]: also, couldnt find the caller of this method

File: e2e/test_sanity.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+import pytest
+
+from kfp_server_api.exceptions import ApiException as KFPApiException
+from kubernetes.client.exceptions import ApiException as K8sApiException
+
+from e2e.constants import GENERIC_KUSTOMIZE_MANIFEST_PATH, DEFAULT_USER_NAMESPACE, CUSTOM_RESOURCE_TEMPLATES_FOLDER
+from e2e.utils import safe_open, wait_for, rand_name
+from e2e.config import metadata, region
+from e2e.cluster import cluster
```
Comment (surajkota) [inline e2e/test_sanity.py:15]: ah! I see how those fixtures are being called now. It was hard to see, usually you wouldn't expect a import to be doing this

thinking how to make this obvious and more customizable ...

File: e2e/clients.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+import subprocess
+import time
+import requests
+
+import pytest
+
+from kubernetes import client, config
+import kfp
+
+def client_from_config(cluster, region):
```
[the remark this answers is not in the document]
Comment (rrrkharse) [reply e2e/clients.py:14]: It reads the cluster config from the k8s context. It's equivalent to:
```
kubectl config use-context <context>
```

i was leaving the door open for the multiple cluster, multiple concurrent test run case where the default context could be changing across runs otherwise.

File: e2e/config.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+    the metadata.
+
+    If a resource is present in the metadata it will not be created.
+
+    If the `--keepsuccess` flag is specified, successfully created resources (e.g. those that
+    did not raise an exception) will not be deleted.
+    """
+    successful_creation = False
+    
+    def delete():
```
[the remark this answers is not in the document]
Comment (rrrkharse) [reply e2e/config.py:89]: not sure i follow, if you could provide an example of having a class to model this that would be great

i guess if I implemented it that way I would be creating a new metadata object for every new resource, which would make it hard to wire through all the fixtures

the metadata class also just stores metadata or resources and doesn't deal with creating or deleting them

basically I need to tell pytest to run some function on delete (as part of the finalizer) so I thought it would be better to model create and delete as functions and pass them in

File: e2e/test_sanity.py (hunk 7 above)
Comment (akartsky) [reply e2e/test_sanity.py:15]: so the import calls the fixture and creates the cluster ? ya not readable

Comment (rrrkharse) [reply e2e/test_sanity.py:15]: the import imports the fixture

the cluster is created by pytest when the fixture is provided as an argument to the test function

if a fixture has other fixtures as dependencies, then those will also be created provided the fixture dependency is imported

For example

```
################# fixtures.py #################

@pytest.fixture
def cluster():
    return cluster

def kustomize(cluster):
    return kustomize

################# test_file.py ####################
from fixtures import cluster, kustomize

def test(kustomize):
    # pytest wants to create kustomize, but sees kustomize needs a cluster
    # pytest creates the cluster (using the cluster fixture that was imported), then creates kustomize

    # ... test logic
```

**Comment by rrrkharse** [comment]
Ready for review

File: e2e/clients.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (surajkota) [reply e2e/clients.py]: looks ok for now, can be reassesed later

File: e2e/requirements.txt [hunk 10]
```diff
@@ -0,0 +1,65 @@
+absl-py==0.11.0
```
[the remark this answers is not in the document]
Comment (rrrkharse) [reply e2e/requirements.txt]: done

File: e2e/config.py [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+            return None
+
+        return self.params[key]
+
+    def to_file(self):
+        filename = 'metadata-' + str(time.time_ns())
+        filepath = os.path.abspath(os.path.join(METADATA_FOLDER, filename))
+
+        with safe_open(filepath, 'w') as file:
+            json.dump(self.params, file)
```
[the remark this answers is not in the document]
Comment (rrrkharse) [reply e2e/config.py]: I think with json its clear how the python object is modeled, with yaml it's not as clear

I also don't think this file will be touched very much

File: e2e/cluster.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (rrrkharse) [reply e2e/cluster.py]: Marking this as todo fast follow since testing cluster creation takes a while to test and this should not be a blocker with the current configuration

File: e2e/clients.py [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+@pytest.fixture(scope="class")
+def k8s_core_api_client(cluster, region):
+    """
+    API client for interacting with k8s core API, e.g. describe_pods, etc.
+    """
+
+    return client.CoreV1Api(api_client=client_from_config(cluster, region))
+
+@pytest.fixture(scope="class")
+def k8s_custom_objects_api_client(cluster, region):
```
[the remark this answers is not in the document]
Comment (rrrkharse) [reply e2e/clients.py]: done

File: e2e/config.py [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+            return Metadata(json.load(file))
+
+@pytest.fixture(scope="class")
+def metadata(request):
+    """
+    If `--metadata` argument is present, reads the metadata that is passed in.
+    Else, created an empty metadata object.
+    """
+
+    metadata_file = request.config.getoption("--metadata")
```
[the remark this answers is not in the document]
Comment (rrrkharse) [reply e2e/config.py]: done

**Comment by rrrkharse** [comment]
Ready for review

File: distributions/aws/test/e2e/README.md [hunk 14]
```diff
@@ -0,0 +1,51 @@
+
+## Usage
+
+Install dependencies
+```
+python3
```
Comment (mbaijal) [inline distributions/aws/test/e2e/README.md:6]: We should eventually have a one line installation for these using either a script or a conda or pip package as possible.

File: distributions/aws/test/e2e/README.md [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+kustomize   # MUST BE VERSION 3.2.0
+eksctl
+kubectl
+
+awscli # configure your default credentials
+
+# install python dependencies
+pip install -r requirements.txt
```
Comment (mbaijal) [inline distributions/aws/test/e2e/README.md:14]: Can we include black here and also find a way to make the code formatting for these files automated ?

File: distributions/aws/test/e2e/tests/test_sanity.py [hunk 16]
```diff
@@ -0,0 +1,127 @@
+"""
+Installs the vanilla distribution of kubeflow and validates the installation
```
Comment (mbaijal) [inline distributions/aws/test/e2e/tests/test_sanity.py:2]: Can you expand this to describe how we are validating

File: distributions/aws/test/e2e/tests/test_sanity.py [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+
+    def test_run_pipeline(self, setup, kfp_client):
+        experiment_name = rand_name('experiment-')
+        experiment_description = rand_name('description-')
+        experiment = kfp_client.create_experiment(experiment_name, description=experiment_description, namespace=DEFAULT_USER_NAMESPACE)
+
+        pipeline_id = kfp_client.get_pipeline_id(PIPELINE_NAME)
+        job_name = rand_name('run-')
+
+        run = kfp_client.run_pipeline(experiment.id, job_name=job_name, pipeline_id=pipeline_id)
```
Comment (mbaijal) [inline distributions/aws/test/e2e/tests/test_sanity.py:84]: Is cleanup missing here for the KF pipeline and associated run and experiment ?

**Review by mbaijal — COMMENTED** [review]
Minor Comments

File: distributions/aws/test/e2e/tests/test_sanity.py (hunk 17 above)
Comment (rrrkharse) [reply distributions/aws/test/e2e/tests/test_sanity.py]: Missed cleanup for experiment. Pipeline runs can't be deleted (https://kubeflow-pipelines.readthedocs.io/en/latest/source/kfp.client.html).

File: distributions/aws/test/e2e/tests/test_sanity.py (hunk 16 above)
Comment (rrrkharse) [reply distributions/aws/test/e2e/tests/test_sanity.py]: done

File: distributions/aws/test/e2e/README.md (hunk 15 above)
Comment (rrrkharse) [reply distributions/aws/test/e2e/README.md:14]: done

**Comment by rrrkharse** [comment]
Ready for review

**Review by mbaijal — APPROVED** [review]
Looks good!

**Review by surajkota — APPROVED** [review]
